// @earlytexts/markit/dist/index.js
// src/compile.ts
import { readFileSync } from "node:fs";

// src/compile/makeError.ts
var makeError_default = ({
  message,
  line,
  length,
  column = 0,
  lines = 0,
  severity = "error"
}) => ({
  message,
  line: line + 1,
  // Convert to 1-based line number
  column: column + 1,
  // Convert to 1-based column number
  endLine: line + lines + 1,
  endColumn: column + length + 1,
  severity
});

// src/compile/compileExternalChildren.ts
var compileExternalChildren_default = (treeWithMetadata, options, loadingStack, compile2) => {
  const childrenPaths = treeWithMetadata.metadata.children;
  if (!childrenPaths) {
    return [[], []];
  }
  const childrenPositions = treeWithMetadata.metadataPositions.children;
  if (!Array.isArray(childrenPaths)) {
    const error = makeError_default({
      message: "The 'children' metadata field must be an array of strings (file paths)",
      line: childrenPositions.line,
      column: childrenPositions.column,
      length: "children".length
    });
    return [[], [error]];
  }
  const errors = [];
  const externalChildren = [];
  for (let i = 0; i < childrenPaths.length; i++) {
    const childPath = childrenPaths[i];
    const elementPosition = childrenPositions.arrayElementPositions[i];
    if (typeof childPath !== "string") {
      errors.push(
        makeError_default({
          message: "Each item in 'children' metadata array must be a string (file path)",
          line: elementPosition.line,
          column: elementPosition.column,
          length: elementPosition.length
        })
      );
      continue;
    }
    const resolvedPath = resolvePath(options.filePath, childPath);
    const normalizedPath = normalizePath(resolvedPath);
    if (loadingStack.has(normalizedPath)) {
      errors.push(
        makeError_default({
          message: "Circular dependency detected",
          line: elementPosition.line,
          column: elementPosition.column,
          length: elementPosition.length
        })
      );
      continue;
    } else {
      loadingStack.add(normalizedPath);
    }
    let fileContent = null;
    let actualPath = resolvedPath;
    try {
      fileContent = options.fileLoader(resolvedPath);
    } catch {
      const pathWithExtension = `${resolvedPath}.mit`;
      try {
        fileContent = options.fileLoader(pathWithExtension);
        actualPath = pathWithExtension;
      } catch {
        const indexPath = `${resolvedPath}/index.mit`;
        try {
          fileContent = options.fileLoader(indexPath);
          actualPath = indexPath;
        } catch (error) {
          errors.push(
            makeError_default({
              message: `Cannot load external child: ${childPath}`,
              line: elementPosition.line,
              column: elementPosition.column,
              length: elementPosition.length
            })
          );
          continue;
        }
      }
    }
    const [childDocument, childErrors] = compile2(
      fileContent,
      {
        ...options,
        filePath: actualPath
      },
      loadingStack
    );
    externalChildren.push(childDocument);
    errors.push(
      ...childErrors.filter((e) => e.severity !== "warning").map((error) => ({ ...error, file: childPath }))
    );
    if (childErrors.length > 0) {
      errors.push(
        makeError_default({
          message: `External child '${childPath}' has errors`,
          line: elementPosition.line,
          column: elementPosition.column,
          length: elementPosition.length,
          severity: "warning"
        })
      );
    }
  }
  return [externalChildren, errors];
};
var resolvePath = (parentPath, relativePath) => {
  if (relativePath.startsWith("/") || relativePath.startsWith("\\")) {
    return relativePath;
  }
  const lastSlash = Math.max(
    parentPath.lastIndexOf("/"),
    parentPath.lastIndexOf("\\")
  );
  if (lastSlash < 0) {
    return relativePath;
  }
  const parentDir = parentPath.substring(0, lastSlash + 1);
  return `${parentDir}${relativePath}`;
};
var normalizePath = (path) => {
  return path.replace(/\\/g, "/").toLowerCase();
};

// src/compile/generateTextTree.ts
var generateTextTree_default = (blocks) => {
  const errors = [];
  const rootLine = blocks[0].lines[0];
  const [rootLevel, rootId] = readTextHeading(rootLine.content);
  if (rootLevel !== 1) {
    const message = rootLevel === 0 ? "Document must begin with a level 1 header (e.g. # Document.Id)" : `Expected level 1 header but found level ${rootLevel}`;
    errors.push(
      makeError_default({
        message,
        line: 0,
        length: rootLine.content.length
      })
    );
  }
  const rootText = {
    id: rootId,
    level: rootLevel,
    startLine: blocks[0].startLine,
    endLine: blocks[0].endLine,
    blocks: [],
    children: []
  };
  let currentLevel = rootLevel;
  const flatTexts = [rootText];
  blocks.slice(1).forEach((block) => {
    const firstLine = block.lines[0];
    const [level, id] = readTextHeading(firstLine.content);
    if (level > 0) {
      if (level > currentLevel + 1) {
        errors.push(
          makeError_default({
            message: `Level ${level} header cannot follow level ${currentLevel} header without an intermediate level`,
            line: block.startLine,
            length: firstLine.content.length
          })
        );
      }
      flatTexts.push({
        level,
        id,
        startLine: block.startLine,
        endLine: block.endLine,
        blocks: [],
        children: []
      });
      currentLevel = level;
    } else {
      const currentText = flatTexts.at(-1);
      currentText.blocks.push(block);
      currentText.endLine = block.endLine;
    }
  });
  flatTexts.slice(1).forEach((text, index) => {
    const parent = flatTexts.slice(0, index + 1).reverse().find((t) => t.level < text.level);
    parent.children.push(text);
  });
  fixEndLines(rootText);
  return [rootText, errors];
};
var readTextHeading = (line) => {
  const idMatch = line.trim().match(/^(#+)\s*([^\s#{}]+)/);
  const level = idMatch ? idMatch[1].length : 0;
  const id = idMatch ? idMatch[2] : "missing-id";
  return [level, id];
};
var fixEndLines = (text) => {
  if (text.children.length > 0) {
    text.children.forEach(fixEndLines);
    text.endLine = text.children.at(-1).endLine;
  }
};

// src/types.ts
var RESERVED_TEXT_KEYS = ["id", "blocks", "children"];
var RESERVED_BLOCK_KEYS = ["id", "content"];
var startLine = /* @__PURE__ */ Symbol("startLine");
var endLine = /* @__PURE__ */ Symbol("endLine");
var leafElements = [
  { trigger: "~~", type: "emSpace" },
  { trigger: "~", type: "nbSpace" },
  { trigger: "//", type: "lineBreak" },
  { trigger: "|", type: "pageBreak" }
];
var headingSpec = {
  marker: "\xA3",
  minLevel: 1,
  maxLevel: 6,
  blockLevel: true
};
var footnoteReferenceSpec = {
  open: "<",
  close: ">",
  pattern: /^n[^\s#{}]+$/,
  type: "footnoteReference"
};
var wrapperElements = [
  { open: '""', close: '""', type: "blockquote" },
  { open: '"', close: '"', type: "quote" },
  { open: "*", close: "*", type: "strong" },
  { open: "_", close: "_", type: "emphasis" },
  { open: "$$", close: "$$", type: "greek" },
  { open: "$", close: "$", type: "foreign" },
  { open: "@", close: "@", type: "aside" },
  { open: "++", close: "++", type: "insertion" },
  { open: "--", close: "--", type: "deletion" },
  { open: "[", close: "]", type: "citation" }
];
var braceCodes = [
  { code: "SS", result: "\xA7" },
  { code: "ae", result: "\xE6" },
  { code: "AE", result: "\xC6" },
  { code: "oe", result: "\u0153" },
  { code: "OE", result: "\u0152" },
  { code: "-", result: "\u2013" },
  { code: "--", result: "\u2014" }
];
var isWrapperElement = (element) => wrapperElements.some((wrapper) => wrapper.type === element.type);
var isBlockLevelType = (type) => type === "heading" || type === "blockquote" || type === "lineBreak";

// src/compile/buildPositionMap.ts
var buildPositionMap_default = (block) => {
  const map = [];
  const originalLineCount = block.endLine - block.startLine + 1;
  const contentLineCount = block.lines.length;
  const tagOnOwnLine = contentLineCount < originalLineCount;
  const lineOffset = tagOnOwnLine ? 1 : 0;
  block.lines.forEach((line, lineIndex) => {
    const actualLine = block.startLine + lineOffset + lineIndex;
    for (let i = 0; i < line.content.length; i++) {
      map.push({
        line: actualLine,
        column: line.charOffset + i
      });
    }
    if (lineIndex < block.lines.length - 1) {
      map.push({
        line: actualLine,
        column: line.charOffset + line.content.length
      });
    }
  });
  return map;
};

// src/compile/transliterateGreek.ts
var transliterateGreek_default = (content) => content.map(transliterateElement);
var transliterateElement = (element) => {
  if (element.type === "plainText") {
    return { ...element, content: transliterateContent(element.content) };
  } else if ("content" in element && Array.isArray(element.content)) {
    return { ...element, content: element.content.map(transliterateElement) };
  }
  return element;
};
var transliterateContent = (input) => {
  let result = "";
  let pos = 0;
  while (pos < input.length) {
    const digraph = digraphs.find(([latin]) => input.startsWith(latin, pos));
    if (digraph) {
      result += digraph[1];
      pos += digraph[0].length;
      continue;
    }
    const char = input[pos];
    const lower = lowerMap[char];
    const upper = upperMap[char];
    if (lower) {
      if (char === "s" && isWordBoundary(input[pos + 1])) {
        result += "\u03C2";
      } else {
        result += lower;
      }
      pos += 1;
    } else if (upper) {
      result += upper;
      pos += 1;
    } else {
      result += char;
      pos += 1;
    }
  }
  return result;
};
var isWordBoundary = (char) => {
  return char === void 0 || /\s/.test(char) || /[.,;:!?'"()[\]{}<>\/\\]/.test(char);
};
var digraphs = [
  ["th", "\u03B8"],
  ["Th", "\u0398"],
  ["TH", "\u0398"],
  ["ph", "\u03C6"],
  ["Ph", "\u03A6"],
  ["PH", "\u03A6"],
  ["ch", "\u03C7"],
  ["Ch", "\u03A7"],
  ["CH", "\u03A7"],
  ["ps", "\u03C8"],
  ["Ps", "\u03A8"],
  ["PS", "\u03A8"]
];
var lowerMap = {
  a: "\u03B1",
  b: "\u03B2",
  g: "\u03B3",
  d: "\u03B4",
  e: "\u03B5",
  z: "\u03B6",
  i: "\u03B9",
  k: "\u03BA",
  l: "\u03BB",
  m: "\u03BC",
  n: "\u03BD",
  x: "\u03BE",
  o: "\u03BF",
  p: "\u03C0",
  r: "\u03C1",
  s: "\u03C3",
  t: "\u03C4",
  u: "\u03C5",
  y: "\u03C5",
  w: "\u03C9",
  h: "\u03B7"
};
var upperMap = {
  A: "\u0391",
  B: "\u0392",
  G: "\u0393",
  D: "\u0394",
  E: "\u0395",
  Z: "\u0396",
  I: "\u0399",
  K: "\u039A",
  L: "\u039B",
  M: "\u039C",
  N: "\u039D",
  X: "\u039E",
  O: "\u039F",
  P: "\u03A0",
  R: "\u03A1",
  S: "\u03A3",
  T: "\u03A4",
  U: "\u03A5",
  Y: "\u03A5",
  W: "\u03A9",
  H: "\u0397"
};

// src/compile/parseElements.ts
var parseElements_default = (input, positionMap, footnoteIds) => {
  const errors = [];
  const [elements] = parseElements(
    input,
    0,
    null,
    false,
    positionMap,
    footnoteIds,
    errors
  );
  const cleanedElements = cleanupElements(elements);
  return [cleanedElements, errors];
};
var parseElements = (input, startPos, closeMarker, insideBlockLevel, positionMap, footnoteIds, errors) => {
  const result = [];
  let pos = startPos;
  let plainTextBuffer = "";
  const flushPlainText = () => {
    if (plainTextBuffer.length > 0) {
      result.push({ type: "plainText", content: plainTextBuffer });
      plainTextBuffer = "";
    }
  };
  while (pos < input.length) {
    if (closeMarker && input.startsWith(closeMarker, pos)) {
      flushPlainText();
      return [result, pos + closeMarker.length];
    }
    if (input[pos] === "\\") {
      if (pos + 1 < input.length) {
        plainTextBuffer += input[pos + 1];
        pos += 2;
        continue;
      } else {
        plainTextBuffer += "\\";
        pos++;
        continue;
      }
    }
    if (input[pos] === "{") {
      const closeBracePos = input.indexOf("}", pos + 1);
      if (closeBracePos === -1) {
        const position = positionMap[pos];
        errors.push(
          makeError_default({
            message: "Unclosed brace code",
            line: position.line,
            column: position.column,
            length: 1
          })
        );
        plainTextBuffer += input[pos];
        pos++;
        continue;
      }
      const code = input.slice(pos + 1, closeBracePos);
      const braceCode = braceCodes.find((bc) => bc.code === code);
      if (braceCode) {
        flushPlainText();
        result.push({ type: "plainText", content: braceCode.result });
        pos = closeBracePos + 1;
        continue;
      } else {
        const position = positionMap[pos + 1];
        errors.push(
          makeError_default({
            message: `Unknown brace code: ${code}`,
            line: position.line,
            column: position.column,
            length: code.length
          })
        );
        plainTextBuffer += input.slice(pos, closeBracePos + 1);
        pos = closeBracePos + 1;
        continue;
      }
    }
    let leafMatched = false;
    for (const leaf of [...leafElements].sort(
      (a, b) => b.trigger.length - a.trigger.length
    )) {
      if (input.startsWith(leaf.trigger, pos)) {
        flushPlainText();
        result.push({ type: leaf.type });
        pos += leaf.trigger.length;
        leafMatched = true;
        break;
      }
    }
    if (leafMatched) continue;
    if (input[pos] === "<") {
      const closeAnglePos = input.indexOf(">", pos + 1);
      if (closeAnglePos !== -1) {
        const refId = input.slice(pos + 1, closeAnglePos);
        if (footnoteReferenceSpec.pattern.test(refId)) {
          flushPlainText();
          result.push({ type: "footnoteReference", id: refId });
          if (!footnoteIds.includes(refId)) {
            const position = positionMap[pos];
            errors.push(
              makeError_default({
                message: `Footnote not found: ${refId}`,
                line: position.line,
                column: position.column,
                length: closeAnglePos - pos + 1
              })
            );
          }
          pos = closeAnglePos + 1;
          continue;
        }
      }
    }
    if (input[pos] === headingSpec.marker) {
      const levelChar = input[pos + 1];
      if (levelChar && /[1-6]/.test(levelChar)) {
        const level = parseInt(levelChar, 10);
        const hasSpace = input[pos + 2] === " ";
        if (hasSpace) {
          if (insideBlockLevel) {
            const position = positionMap[pos];
            errors.push(
              makeError_default({
                message: "Block-level elements cannot be nested",
                line: position.line,
                column: position.column,
                length: 3
              })
            );
            plainTextBuffer += input.slice(pos, pos + 3);
            pos += 3;
            continue;
          }
          const closeMarkerStr = `${headingSpec.marker}${level}`;
          const [headingContent, newPos] = parseElements(
            input,
            pos + 3,
            closeMarkerStr,
            true,
            positionMap,
            footnoteIds,
            errors
          );
          if (newPos === pos + 3 || !input.startsWith(closeMarkerStr, newPos - closeMarkerStr.length)) {
            const position = positionMap[pos];
            errors.push(
              makeError_default({
                message: `Unclosed heading level ${level}`,
                line: position.line,
                column: position.column,
                length: 3
              })
            );
          }
          flushPlainText();
          result.push({ type: "heading", level, content: headingContent });
          pos = newPos;
          continue;
        }
      }
    }
    let wrapperMatched = false;
    for (const wrapper of [...wrapperElements].sort(
      (a, b) => b.open.length - a.open.length
    )) {
      if (input.startsWith(wrapper.open, pos)) {
        if (isBlockLevelType(wrapper.type) && insideBlockLevel) {
          const position = positionMap[pos];
          const closeIdx = input.indexOf(
            wrapper.close,
            pos + wrapper.open.length
          );
          const endPos = closeIdx >= 0 ? closeIdx + wrapper.close.length : pos + wrapper.open.length;
          errors.push(
            makeError_default({
              message: "Block-level elements cannot be nested",
              line: position.line,
              column: position.column,
              length: endPos - pos
            })
          );
          plainTextBuffer += input.slice(pos, endPos);
          pos = endPos;
          wrapperMatched = true;
          break;
        }
        const [wrapperContent, newPos] = parseElements(
          input,
          pos + wrapper.open.length,
          wrapper.close,
          isBlockLevelType(wrapper.type) || insideBlockLevel,
          positionMap,
          footnoteIds,
          errors
        );
        if (newPos === pos + wrapper.open.length || !input.startsWith(wrapper.close, newPos - wrapper.close.length)) {
          const position = positionMap[pos];
          errors.push(
            makeError_default({
              message: `Unclosed formatting: ${wrapper.open}`,
              line: position.line,
              column: position.column,
              length: wrapper.open.length
            })
          );
        }
        flushPlainText();
        const finalContent = wrapper.type === "greek" ? transliterateGreek_default(wrapperContent) : wrapperContent;
        result.push({ type: wrapper.type, content: finalContent });
        pos = newPos;
        wrapperMatched = true;
        break;
      }
    }
    if (wrapperMatched) continue;
    plainTextBuffer += input[pos];
    pos++;
  }
  if (closeMarker && pos >= input.length) {
  }
  flushPlainText();
  return [result, pos];
};
var cleanupElements = (elements) => {
  const result = [];
  for (let i = 0; i < elements.length; i++) {
    const element = elements[i];
    if (element.type === "plainText" && /^\s+$/.test(element.content)) {
      const prevElement = result[result.length - 1];
      const nextElement = elements[i + 1];
      if (!nextElement || prevElement && isBlockLevelType(prevElement.type) || nextElement && isBlockLevelType(nextElement.type)) {
        continue;
      }
    }
    if (element.type === "plainText") {
      const nextElement = elements[i + 1];
      if (nextElement && isBlockLevelType(nextElement.type)) {
        result.push({
          type: "plainText",
          content: element.content.trimEnd()
        });
        continue;
      }
    }
    if (element.type === "plainText") {
      const prevElement = result[result.length - 1];
      if (prevElement && isBlockLevelType(prevElement.type)) {
        result.push({
          type: "plainText",
          content: element.content.trimStart()
        });
        continue;
      }
    }
    if (element.type === "heading" || isWrapperElement(element)) {
      const cleanedContent = cleanupElements(element.content);
      if (isBlockLevelType(element.type)) {
        while (cleanedContent.length > 0 && cleanedContent[0].type === "plainText" && typeof cleanedContent[0].content === "string" && /^\s+$/.test(cleanedContent[0].content)) {
          cleanedContent.shift();
        }
        const firstElement = cleanedContent[0];
        if (firstElement && firstElement.type === "plainText" && typeof firstElement.content === "string") {
          cleanedContent[0] = {
            type: "plainText",
            content: firstElement.content.trimStart()
          };
        }
        const lastIndex = cleanedContent.length - 1;
        const lastElement = cleanedContent[lastIndex];
        if (lastElement && lastElement.type === "plainText" && typeof lastElement.content === "string") {
          cleanedContent[lastIndex] = {
            type: "plainText",
            content: lastElement.content.trimEnd()
          };
        }
      }
      result.push({ ...element, content: cleanedContent });
    } else {
      result.push(element);
    }
  }
  return result;
};

// src/compile/parseContent.ts
var parseContent_default = (tree, externalChildren = []) => {
  return parseTextContent(tree, externalChildren, {});
};
var parseTextContent = (text, externalChildren = [], parentMetadata) => {
  const footnoteIds = text.blocks.filter((b) => footnoteReferenceSpec.pattern.test(b.id)).map((b) => b.id);
  const blockResults = text.blocks.map(
    (block) => parseBlockContent(block, footnoteIds)
  );
  const blocks = blockResults.map((result) => result[0]);
  const blockErrors = blockResults.flatMap((result) => result[1]);
  const mergedMetadata = { ...parentMetadata, ...text.metadata };
  const childResults = text.children.map(
    (child) => parseTextContent(child, [], mergedMetadata)
  );
  const children = childResults.map((result) => result[0]);
  const childErrors = childResults.flatMap((result) => result[1]);
  const document = {
    ...mergedMetadata,
    id: text.id,
    blocks,
    children: [...children, ...externalChildren],
    [startLine]: text.startLine,
    [endLine]: text.endLine
  };
  return [document, [...blockErrors, ...childErrors]];
};
var parseBlockContent = (block, footnoteIds) => {
  const text = block.lines.map((line) => line.content).join(" ").replace(/\s+/g, " ").trim();
  const positionMap = buildPositionMap_default(block);
  const [content, errors] = parseElements_default(text, positionMap, footnoteIds);
  const parsedBlock = {
    ...block.metadata,
    id: block.id,
    content,
    [startLine]: block.startLine,
    [endLine]: block.endLine
  };
  return [parsedBlock, errors];
};

// src/compile/parseMetadata.ts
var parseMetadata_default = (textTree) => {
  return parseTextMetadata(textTree);
};
var parseTextMetadata = (text) => {
  const firstBlock = text.blocks[0];
  const firstLine = firstBlock?.lines[0];
  const isMetadata = firstLine?.content.match(/^\w+:/);
  const [metadata, metadataPositions, metadataErrors] = isMetadata ? parseMetadataBlock(firstBlock) : [{}, {}, []];
  const contentBlocks = isMetadata ? text.blocks.slice(1) : text.blocks;
  const parseBlockMetadataResult = contentBlocks.reduce(
    (acc, block) => {
      const [blockWithMetadata, blockErrors2] = parseBlockMetadata(
        block,
        acc.map((b) => b[0])
      );
      acc.push([blockWithMetadata, blockErrors2]);
      return acc;
    },
    []
  );
  const blocksWithMetadata = parseBlockMetadataResult.map(
    (result) => result[0]
  );
  const blockErrors = parseBlockMetadataResult.flatMap((result) => result[1]);
  const footnoteErrors = [];
  let firstFootnoteIndex = null;
  for (let i = 0; i < blocksWithMetadata.length; i++) {
    const isFootnote = footnoteReferenceSpec.pattern.test(
      blocksWithMetadata[i].id
    );
    if (isFootnote && firstFootnoteIndex === null) {
      firstFootnoteIndex = i;
    } else if (!isFootnote && firstFootnoteIndex !== null) {
      const footnoteBlock = blocksWithMetadata[firstFootnoteIndex];
      const rawFootnoteBlock = contentBlocks[firstFootnoteIndex];
      footnoteErrors.push(
        makeError_default({
          message: "Footnote blocks must appear after all paragraph blocks",
          line: footnoteBlock.startLine,
          column: rawFootnoteBlock.lines[0].charOffset,
          length: footnoteBlock.id.length + 3
          // {# + id + }
        })
      );
      break;
    }
  }
  const parseChildrenResult = text.children.map(parseTextMetadata);
  const childrenWithMetadata = parseChildrenResult.map((result) => result[0]);
  const childrenErrors = parseChildrenResult.flatMap((result) => result[1]);
  const textWithMetadata = {
    ...text,
    metadata,
    metadataPositions,
    blocks: blocksWithMetadata,
    children: childrenWithMetadata
  };
  const errors = [
    ...metadataErrors,
    ...blockErrors,
    ...footnoteErrors,
    ...childrenErrors
  ];
  return [textWithMetadata, errors];
};
var parseMetadataBlock = (block) => {
  const errors = [];
  const metadata = {};
  const metadataPositions = {};
  for (let index = 0; index < block.lines.length; index++) {
    const line = block.lines[index];
    const multilineArrayMatch = line.content.match(/^(\w+)\s*:\s*$/);
    if (multilineArrayMatch) {
      const key2 = multilineArrayMatch[1];
      metadataPositions[key2] = {
        line: block.startLine + index,
        column: line.charOffset,
        length: line.content.length,
        arrayElementPositions: []
      };
      const arrayItems = [];
      let arrayIndex = index + 1;
      while (arrayIndex < block.lines.length) {
        const arrayLine = block.lines[arrayIndex];
        const arrayItemMatch = arrayLine.content.match(/^- (.+)$/);
        if (!arrayItemMatch) {
          break;
        }
        const itemString = arrayItemMatch[1].trim();
        const itemStartColumn = arrayLine.charOffset + arrayLine.content.indexOf(itemString);
        metadataPositions[key2].arrayElementPositions.push(
          {
            line: block.startLine + arrayIndex,
            column: itemStartColumn,
            length: itemString.length
          }
        );
        let itemValue;
        try {
          itemValue = JSON.parse(itemString);
        } catch {
          itemValue = itemString;
          errors.push(
            makeError_default({
              message: `Invalid metadata value: ${itemString}`,
              line: block.startLine + arrayIndex,
              column: itemStartColumn,
              length: itemString.length
            })
          );
        }
        arrayItems.push(itemValue);
        arrayIndex++;
      }
      if (arrayItems.length === 0) {
        errors.push(
          makeError_default({
            message: "Multiline array must have at least one item",
            line: block.startLine + index,
            column: line.charOffset,
            length: line.content.length
          })
        );
      } else {
        const types = new Set(arrayItems.map((item) => typeof item));
        if (types.size > 1) {
          errors.push(
            makeError_default({
              message: "Array contains mixed types (arrays must contain only numbers, only booleans, or only strings)",
              line: block.startLine + index,
              column: line.charOffset,
              length: line.content.length
            })
          );
        }
        metadata[key2] = arrayItems;
      }
      index = arrayIndex - 1;
      continue;
    }
    const match = line.content.match(/^(\w+)\s*:\s*(.+)$/);
    if (!match) {
      errors.push(
        makeError_default({
          message: "Invalid metadata line, expected 'key: value'",
          line: block.startLine + index,
          column: line.charOffset,
          length: line.content.length
        })
      );
      continue;
    }
    const key = match[1];
    const valueString = match[2].trim();
    if (RESERVED_TEXT_KEYS.includes(key) && key !== "children") {
      errors.push(
        makeError_default({
          message: `The '${key}' metadata key is reserved and cannot be used in the document metadata`,
          line: block.startLine + index,
          column: line.charOffset,
          length: key.length
        })
      );
      continue;
    }
    metadataPositions[key] = {
      line: block.startLine + index,
      column: line.charOffset,
      length: line.content.length,
      arrayElementPositions: []
    };
    let value;
    try {
      value = JSON.parse(valueString);
    } catch {
      value = valueString;
      errors.push(
        makeError_default({
          message: `Invalid metadata value: ${valueString}`,
          line: block.startLine + index,
          column: line.charOffset + line.content.indexOf(valueString),
          length: valueString.length
        })
      );
    }
    if (Array.isArray(value)) {
      const arrayOpeningBracketIndex = line.content.indexOf(valueString);
      value.forEach((item) => {
        const itemString = JSON.stringify(item);
        const itemStartIndex = line.content.indexOf(
          itemString,
          arrayOpeningBracketIndex
        );
        metadataPositions[key].arrayElementPositions.push({
          line: block.startLine + index,
          column: line.charOffset + itemStartIndex,
          length: itemString.length
        });
      });
    }
    if (Array.isArray(value)) {
      const types = new Set(value.map((item) => typeof item));
      if (types.size > 1) {
        errors.push(
          makeError_default({
            message: "Array contains mixed types (arrays must contain only numbers, only booleans, or only strings)",
            line: block.startLine + index,
            column: line.charOffset,
            length: line.content.length
          })
        );
      }
    }
    metadata[key] = value;
  }
  return [metadata, metadataPositions, errors];
};
var parseBlockMetadata = (block, previousBlocks) => {
  const errors = [];
  const [firstLine, ...otherLines] = block.lines;
  const blockTagMatch = firstLine.content.match(/^\{#(.+?)\}/);
  if (!blockTagMatch) {
    const message = firstLine.content.trim().startsWith("{#") ? "Block tag is not properly closed with '}'" : "Block is missing metadata tag '{#id}'";
    errors.push(
      makeError_default({
        message,
        line: block.startLine,
        column: firstLine.charOffset,
        length: firstLine.content.length
      })
    );
  }
  const blockTagContent = blockTagMatch ? blockTagMatch[1].trim() : `${block.startLine}`;
  const blockTagParts = blockTagContent.split(",").map((part) => part.trim());
  const id = blockTagParts[0];
  if (blockTagMatch && !/^[^\s#{}]+$/.test(id)) {
    const idOffset = firstLine.content.indexOf(id, 2);
    errors.push(
      makeError_default({
        message: `Block ID '${id}' contains invalid characters (IDs may not contain whitespace, '#', '{', or '}')`,
        line: block.startLine,
        column: firstLine.charOffset + idOffset,
        length: id.length
      })
    );
  } else if (blockTagMatch && id.startsWith("n") && !footnoteReferenceSpec.pattern.test(id)) {
    const idOffset = firstLine.content.indexOf(id, 2);
    errors.push(
      makeError_default({
        message: `Block ID '${id}' is not a valid footnote ID (footnote IDs must start with 'n' followed by at least one character)`,
        line: block.startLine,
        column: firstLine.charOffset + idOffset,
        length: id.length
      })
    );
  }
  if (previousBlocks.some((b) => b.id === id)) {
    errors.push(
      makeError_default({
        message: `Duplicate block ID: #${id}`,
        line: block.startLine,
        column: firstLine.charOffset,
        length: firstLine.content.length
      })
    );
  }
  const metadata = {};
  blockTagParts.slice(1).forEach((part) => {
    const [key, valueString] = part.split("=").map((s) => s.trim());
    if (!key || !valueString) {
      errors.push(
        makeError_default({
          message: "Invalid block metadata, expected 'key=value'",
          line: block.startLine,
          column: firstLine.charOffset + firstLine.content.indexOf(part),
          length: part.length
        })
      );
      return;
    }
    if (RESERVED_BLOCK_KEYS.includes(key)) {
      errors.push(
        makeError_default({
          message: `Block tag key '${key}' is reserved and cannot be used in metadata`,
          line: block.startLine,
          column: firstLine.charOffset + firstLine.content.indexOf(part),
          length: key.length
        })
      );
      return;
    }
    let value;
    try {
      value = JSON.parse(valueString);
    } catch {
      value = valueString;
      errors.push(
        makeError_default({
          message: `Invalid metadata value: ${valueString}`,
          line: block.startLine,
          column: firstLine.charOffset + firstLine.content.indexOf(valueString),
          length: valueString.length
        })
      );
    }
    metadata[key] = value;
  });
  const contentAfterTag = blockTagMatch ? firstLine.content.slice(blockTagMatch[0].length).trim() : firstLine.content.trim().startsWith("{#") ? "" : firstLine.content.trim();
  const newFirstLine = contentAfterTag ? {
    charOffset: firstLine.charOffset + firstLine.content.indexOf(contentAfterTag),
    content: contentAfterTag
  } : null;
  const lines = newFirstLine ? [newFirstLine, ...otherLines] : otherLines;
  const blockWithMetadata = {
    ...block,
    id,
    metadata,
    lines
  };
  return [blockWithMetadata, errors];
};

// src/compile/splitIntoBlocks.ts
var splitIntoBlocks_default = (text) => {
  const lines = text.split("\n");
  let blankLines = 0;
  const blocks = [];
  lines.forEach((line, index) => {
    const trimmed = line.trim();
    if (trimmed === "") {
      blankLines++;
    } else {
      const lineObject = {
        charOffset: line.indexOf(trimmed),
        content: trimmed
      };
      if (blankLines > 0) {
        blocks.push({ startLine: index, endLine: index, lines: [lineObject] });
        blankLines = 0;
      } else {
        const lastBlock = blocks.at(-1);
        if (!lastBlock) {
          blocks.push({
            startLine: index,
            endLine: index,
            lines: [lineObject]
          });
        } else {
          lastBlock.lines.push(lineObject);
          lastBlock.endLine = index;
        }
      }
    }
  });
  return blocks;
};

// src/compile.ts
var compile_default = (text, options = {}) => {
  const loadingStack = new Set(options.filePath ? [options.filePath] : []);
  return compile(text, options, loadingStack);
};
var compile = (text, optionOverrides, loadingStack = /* @__PURE__ */ new Set()) => {
  const options = {
    embedExternalChildren: true,
    fileLoader: (path) => readFileSync(path, "utf-8"),
    filePath: "",
    ...optionOverrides
  };
  const [firstBlock, ...otherBlocks] = splitIntoBlocks_default(text);
  if (!firstBlock) {
    const emptyDocument = {
      id: "empty-document",
      blocks: [],
      children: [],
      [startLine]: 0,
      [endLine]: 0
    };
    const emptyDocumentError = makeError_default({
      message: "Document is empty",
      line: 0,
      column: 0,
      length: 0
    });
    return [emptyDocument, [emptyDocumentError]];
  }
  const [textTree, treeErrors] = generateTextTree_default([firstBlock, ...otherBlocks]);
  const [treeWithMetadata, metaDataErrors] = parseMetadata_default(textTree);
  const [document, contentErrors] = parseContent_default(treeWithMetadata);
  const [externalChildren, externalChildrenErrors] = options.embedExternalChildren ? compileExternalChildren_default(
    treeWithMetadata,
    options,
    loadingStack,
    compile
  ) : [[], []];
  const fullDocument = {
    ...document,
    children: [...document.children, ...externalChildren],
    [startLine]: document[startLine],
    [endLine]: document[endLine]
  };
  const errors = [
    ...treeErrors,
    ...metaDataErrors,
    ...externalChildrenErrors,
    ...contentErrors
  ].sort((a, b) => a.line - b.line || a.column - b.column);
  return [fullDocument, errors];
};

// src/format/helpers.ts
var emitLine = (state, line) => {
  return {
    ...state,
    acc: [...state.acc, line],
    lastEmitted: "nonblank"
  };
};
var emitBlank = (state) => {
  if (state.lastEmitted === "blank") return state;
  return {
    ...state,
    acc: [...state.acc, ""],
    lastEmitted: "blank"
  };
};
var flushContent = (state) => {
  if (state.contentBuffer.length === 0) return state;
  const collapsedContent = state.contentBuffer.join(" ");
  const extractedLines = extractBlockElements(collapsedContent);
  return {
    ...state,
    acc: [...state.acc, ...extractedLines],
    contentBuffer: [],
    lastEmitted: "nonblank"
  };
};
var extractBlockElements = (content) => {
  const lines = [];
  const blockRegex = /(£\d[^£]*£\d|""[^"]*"")/g;
  let lastIndex = 0;
  let match;
  while ((match = blockRegex.exec(content)) !== null) {
    const beforeBlock = content.slice(lastIndex, match.index).trim();
    const blockElement = match[0];
    if (beforeBlock) {
      lines.push(...splitOnLineBreakMarker(beforeBlock, 0));
    }
    if (blockElement.startsWith('"')) {
      lines.push(...splitOnLineBreakMarker(blockElement, 4));
    } else {
      lines.push(...splitOnLineBreakMarker(blockElement, 0));
    }
    lastIndex = match.index + blockElement.length;
  }
  const afterLast = content.slice(lastIndex).trim();
  if (afterLast) {
    lines.push(...splitOnLineBreakMarker(afterLast, 0));
  }
  return lines;
};
var splitOnLineBreakMarker = (text, offset) => {
  return text.replace(/(\S)\/\//g, "$1 //").split("//").map((part, index, array) => {
    const trimmed = part.trim();
    const withLineBreak = index < array.length - 1 ? `${trimmed} //` : trimmed;
    return " ".repeat(offset) + withLineBreak;
  }).filter((part) => part !== "" && part !== "//");
};

// src/format/formatBlockTag.ts
var formatBlockTag_default = (state, line) => {
  let newState = flushContent(state);
  if (newState.context === "inContent") {
    newState = emitBlank(newState);
  } else if (newState.context === "inMetadata") {
    newState = emitBlank(newState);
    newState = { ...newState, context: "afterMetadata" };
  }
  const { tag, content } = formatBlockTag(line);
  newState = emitLine(newState, tag);
  if (content) {
    newState = {
      ...newState,
      contentBuffer: [content],
      context: "inContent"
    };
  } else {
    newState = { ...newState, context: "inContent" };
  }
  return newState;
};
var formatBlockTag = (line) => {
  const trimmed = line.trim();
  const closingBrace = trimmed.indexOf("}");
  if (closingBrace === -1) {
    return { tag: line, content: "" };
  }
  const inner = trimmed.slice(2, closingBrace);
  const rest = trimmed.slice(closingBrace + 1).trim();
  const parts = [];
  let current = "";
  let inQuote = false;
  let escapeNext = false;
  for (let i = 0; i < inner.length; i++) {
    const char = inner[i];
    if (escapeNext) {
      current += char;
      escapeNext = false;
      continue;
    }
    if (char === "\\") {
      current += char;
      escapeNext = true;
      continue;
    }
    if (char === '"') {
      inQuote = !inQuote;
      current += char;
      continue;
    }
    if (char === "," && !inQuote) {
      parts.push(current);
      current = "";
      continue;
    }
    current += char;
  }
  parts.push(current);
  const formattedParts = parts.map(formatBlockTagPart);
  const formattedTag = `{#${formattedParts.join(", ")}}`;
  return { tag: formattedTag, content: rest };
};
var formatBlockTagPart = (part, index) => {
  const trimmedPart = part.trim();
  if (index === 0) {
    return trimmedPart;
  }
  const eqIndex = trimmedPart.indexOf("=");
  if (eqIndex === -1) {
    return trimmedPart;
  }
  const key = trimmedPart.slice(0, eqIndex).trim();
  const value = trimmedPart.slice(eqIndex + 1).trim();
  return `${key}=${value}`;
};

// src/format/formatIdBlock.ts
var formatIdBlock_default = (state, line) => {
  let newState = flushContent(state);
  if (newState.context !== "start") {
    newState = emitBlank(newState);
  }
  const formattedLine = formatIdBlock(line);
  newState = emitLine(newState, formattedLine);
  newState = emitBlank(newState);
  return { ...newState, context: "afterId" };
};
var formatIdBlock = (line) => {
  const match = line.trim().match(/^(#+)\s+(.+)$/);
  const [, hashes, id] = match;
  return `${hashes} ${id?.trim()}`;
};

// src/format/formatMetadata.ts
var formatMetadata_default = (state, line) => {
  const formattedMetadata = formatMetadata(line);
  const newState = emitLine(state, formattedMetadata);
  return { ...newState, context: "inMetadata" };
};
var formatMetadata = (line) => {
  const leadingSpace = line.match(/^\s*/)?.[0] || "";
  const trimmed = line.trim();
  const colonIndex = trimmed.indexOf(":");
  const key = trimmed.slice(0, colonIndex).trim();
  const value = trimmed.slice(colonIndex + 1).trimStart();
  if (!value) return `${leadingSpace}${key}:`;
  return `${leadingSpace}${key}: ${value}`;
};

// src/format/handleBlankLine.ts
var handleBlankLine_default = (state) => {
  if (state.context === "start") {
    return state;
  }
  let newState = flushContent(state);
  return emitBlank(newState);
};

// src/format/handleContentLine.ts
var handleContentLine_default = (state, line) => {
  if (state.context === "inContent" || state.context === "afterMetadata") {
    return {
      ...state,
      contentBuffer: [...state.contentBuffer, line],
      context: "inContent"
    };
  }
  if (state.context === "afterId") {
    return {
      ...state,
      contentBuffer: [line],
      context: "inContent"
    };
  }
  return emitLine(state, line);
};

// src/format.ts
var format_default = (text) => {
  const lines = text.replace(/\r\n?/g, "\n").split("\n");
  const penultimateState = lines.reduce(formatLine, initialState);
  const { acc: finalLines } = flushContent(penultimateState);
  while (finalLines.length > 0 && finalLines.at(-1) === "") {
    finalLines.pop();
  }
  return finalLines.join("\n") + "\n";
};
var initialState = {
  context: "start",
  acc: [],
  contentBuffer: [],
  lastEmitted: "blank"
};
var formatLine = (state, line) => {
  const normalized = line.trimEnd().replace(/(?<=\S)\s+/g, " ");
  const isBlank = normalized.trim() === "";
  const isId = /^(#+)\s+(.+)$/.test(normalized.trim());
  const isBlockTag = normalized.trim().startsWith("{#");
  const hasColon = normalized.includes(":");
  const startsWithDash = normalized.trim().startsWith("-");
  const isMetadataKeyValue = (state.context === "afterId" || state.context === "inMetadata") && !isBlockTag && !isBlank && !startsWithDash && hasColon;
  if (isBlank) return handleBlankLine_default(state);
  if (isId) return formatIdBlock_default(state, normalized);
  if (isBlockTag) return formatBlockTag_default(state, normalized);
  if (isMetadataKeyValue) return formatMetadata_default(state, normalized);
  return handleContentLine_default(state, normalized);
};

// src/renderHTML.ts
var renderHTML_default = (document) => {
  return `<!doctype html><html><head><meta charset="UTF-8"><title>${document.id}</title></head><body>${documentToHTML(
    document
  )}</body></html>`;
};
var documentToHTML = (document) => {
  const line = document[startLine];
  const blocks = document.blocks.map(blockToHTML).join("");
  const children = document.children.map(documentToHTML).join("");
  return `<section id="${document.id}" data-line="${line}">${blocks}${children.length > 0 ? children : ""}</section>`;
};
var blockToHTML = (block) => {
  const line = block[startLine];
  const innerHTML = block.id.startsWith("n") ? `<sup>${block.id}</sup> ${contentToHTML(block.content)}` : contentToHTML(block.content);
  const outerHTML = `<div data-line="${line}"><p>${innerHTML}</p></div>`;
  return outerHTML.replace(/<p><\/p>/g, "");
};
var contentToHTML = (content) => content.map(elementToHTML).join("");
var elementToHTML = (element) => {
  switch (element.type) {
    case "plainText":
      return element.content.replace(/&/g, "&amp;");
    case "lineBreak":
      return "<br />";
    case "pageBreak":
      return '<span class="page-break">|</span>';
    case "nbSpace":
      return "&nbsp;";
    case "emSpace":
      return "&emsp;";
    case "heading":
      return `</p><h${element.level}>${contentToHTML(
        element.content
      )}</h${element.level}><p>`;
    case "footnoteReference":
      return `<a href="#footnote-${element.id}" id="footnote-ref-${element.id}"><sup>${element.id}</sup></a>`;
    case "strong":
      return `<strong>${contentToHTML(element.content)}</strong>`;
    case "emphasis":
      return `<em>${contentToHTML(element.content)}</em>`;
    case "quote":
      return `<q>${contentToHTML(element.content)}</q>`;
    case "blockquote":
      return `</p><blockquote>${contentToHTML(
        element.content
      )}</blockquote><p>`;
    case "foreign":
      return `<em class="foreign">${contentToHTML(element.content)}</em>`;
    case "greek":
      return `<em class="greek">${contentToHTML(element.content)}</em>`;
    case "aside":
      return `<span class="aside">${contentToHTML(element.content)}</span>`;
    case "insertion":
      return `<ins>${contentToHTML(element.content)}</ins>`;
    case "deletion":
      return `<del>${contentToHTML(element.content)}</del>`;
    case "citation":
      return `<cite>${contentToHTML(element.content)}</cite>`;
    /* v8 ignore next 2 */
    default:
      return element;
  }
};

// src/renderText.ts
var renderText_default = (document) => documentToText(document) + "\n";
var documentToText = (document) => {
  const blocks = document.blocks.map(blockToText).join("\n\n");
  const children = document.children.map(documentToText).join("\n\n");
  return `${blocks}${children ? `

${children}` : ""}`;
};
var blockToText = (block) => block.id.startsWith("n") ? `[^${block.id}]: ${contentToText(block.content)}`.trim() : contentToText(block.content).trim();
var contentToText = (content) => content.map(elementToText).join("");
var elementToText = (element) => {
  switch (element.type) {
    case "plainText":
      return element.content;
    case "lineBreak":
      return "\n";
    case "pageBreak":
      return "|";
    case "nbSpace":
      return " ";
    case "emSpace":
      return "  ";
    case "heading":
      return `${contentToText(element.content)}

`;
    case "footnoteReference":
      return `<${element.id}>`;
    case "strong":
      return contentToText(element.content);
    case "emphasis":
      return contentToText(element.content);
    case "quote":
      return `"${contentToText(element.content)}"`;
    case "blockquote":
      return `

    ${contentToText(element.content)}

`;
    case "foreign":
      return contentToText(element.content);
    case "greek":
      return contentToText(element.content);
    case "aside":
      return "";
    case "insertion":
      return contentToText(element.content);
    case "deletion":
      return "";
    case "citation":
      return `[${contentToText(element.content)}]`;
    /* v8 ignore next 2 */
    default:
      return element;
  }
};
export {
  compile_default as compile,
  endLine,
  format_default as format,
  renderHTML_default as renderHTML,
  renderText_default as renderText,
  startLine
};
